docs/rfcs/20220119_raft_replication_simplification.md

```diff
@@ -0,0 +1,162 @@
+being behind. Since recovery stm is currently used for both recovering the 
+follower (when it is actually behind) and delivering follower data replicated 
+with `leader_ack` consistency the `recovery_stm` implementation contains some 
+optimization to minimize the `leader_ack` latency. 
+
+There are two main optimizations:
+- not stopping recovery when last replicate request consistency level is `leader_ack`
+- triggering next dispatch loop immediately after leader disk append
+
+Those optimizations make handling of recovering follower more complex in 
+other scenarios f.e. `leadership_transfer`. 
+
+Since there is no coordination between the two paths that the follower append 
+entries requests are sent from (i.e. `replicate_stm` and `recovery_stm`) those 
+request may be reordered and cause log truncations and redeliveries which lead 
+to increased latency.
+
+In the proposed solution since we are not longer going to relay on the 
+recovery STM to deliver messages to follower in normal conditions the complicated
+handling of recovery finishing logic will no longer be needed. 
+When follower is up to date with the leader `recovery_stm` is simply not running.
+
+The only place that will require coordination is the time when follower 
+is not longer behind the leader but `replicate_stm` skips dispatching requests 
+to the follower as it is still recovering. This may lead to the situation in 
+which next leader append will happen before `recovery_stm` received follower 
+response for last request i.e. it can not finish recovery as log end already 
+advanced. This case may be handled by coordinating dispatching follower 
+requests from `recovery_stm` and `replicate_stm`. 
+We may introduce last sent offset to 
+`raft::follower_index_metadata` to check if recovery should finish 
+instead of the check being based only on the follower response.
+
+This way the `replicate_stm` will not skip sending a request to the follower if 
+that is the next request follower is supposed to receive, on the other hand the 
+`recovery_stm` will not try to send the same request again but will finish 
+assuming response will be successful.
+
+### Implementation details
+
+To implement simplified solution we will pass the `consistency_level` 
+parameter to `raft::replicate_batcher::replicate` method. 
+The `raft::replicate_entries_stm::apply_on_leader` method will return immediately after 
+successful `leader_append`. Full quorum replication round will be finish 
+after `raft::replicate_entries_stm::wait_for_majority()` returns. 
+Replicate batcher based on cached request `consistency_level` will decide when 
+to release the request, either after `apply_on_leader` finishes or 
+`wait_for_majority` finishes.
+
+The same way as in current implementation the `raft::replicate_entries_stm` will
+release batcher memory units after dispatching follower request. 
+This will allow propagating backpressure for both consistency levels.
```

Comment:
```suggestion
As in current implementation, the `raft::replicate_entries_stm` will
release batcher memory units after dispatching the follower request. 
This will allow propagating backpressure for both consistency levels.
```